Widgets_CDAT/DictWidget/dict.py

```diff
@@ -0,0 +1,213 @@
+
+        # for key, value in self.dict.items():
+        col = QHBoxLayout()
+        r_button = QPushButton()
+        r_button.setText("X")
+        r_button.clicked.connect(self.removeRow(col, self.row_count))
+
+        # create right text edit
+        edit_right = QLineEdit()
+        edit_right.setText(str(value))
+        edit_right.setMinimumWidth(100)
+
+        # set text based off if valid_keys has values
+        if self.valid_keys:
+            edit_left = QComboBox()
+            for i in self.valid_keys:
+                edit_left.addItem(i)
+            index = edit_left.findText(str(key))
+            if index != -1:
+                edit_left.setCurrentIndex(index)
+
+            # assign signals
+            edit_left.currentIndexChanged.connect(self.updateLists(edit_left, edit_right))
+
+            self.cBoxes.append(edit_left)
+
+        else:
+            edit_left = QLineEdit()
+            validator = InputChecker(self.row_count)
+            self.validators.append(validator)
+            edit_left.setValidator(validator)
+            validator.correctInput.connect(self.colorValid(edit_left))
+            validator.inputInvalid.connect(self.colorInvalid(edit_left))
+
+            edit_left.setText(str(key))
+            edit_left.editingFinished.connect(self.updateLists(edit_left, edit_right))
+            self.textEdits.append(edit_left)
+
+        edit_right.editingFinished.connect(self.updateLists(edit_left, edit_right))
+        edit_left.setMinimumWidth(100)
+        col.addWidget(r_button)
+        col.addWidget(edit_left)
+        col.addWidget(edit_right)
+        self.rows.insertLayout(self.row_count, col)
+        self.row_count += 1
+
+
+    def removeRow(self, section, index):
+        def performRemove():
+
+            child = section.takeAt(0)
+            c = 0
+            while child:
+                widget = child.widget()
+                if type(widget) == QComboBox:
+                    self.cBoxes.remove(widget)
+
+                if c == 1 and type(widget) != QComboBox:
+                    self.validators.remove(self.validators[index])
+
+                    for i in range(len(self.validators)):
+                        self.validators[i].setIndex(i)
+
+                    self.textEdits.remove(widget)
+                    self.keys.remove(self.keys[index])
+                    self.values.remove(self.values[index])
+
+                widget.deleteLater()
+                child = section.takeAt(0)
+                c += 1
+            section.deleteLater()
+            self.row_count -= 1
+
+            self.textEdited.emit(dict(zip(self.keys, self.values)))
+
+        return performRemove
+
+
+    # set valid keys to be selected
+    def setValidKeys(self, keys):
+        self.valid_keys = keys
+        self.valid_keys.insert(0, "")
+        self.cBoxes = []
+
+    # set inital dictionary values
+    def setDict(self, dictionary):
+
+        for key, value in dictionary.items():
```

Comment:
You should clear the current screen before inserting more rows; if I call `setDict({"1": "Two", "2": "Three"})` and then `setDict({"3": "Four"})`, I get all three of those options on the screen at once.